### PR TITLE
Use Str instead of Symbol for name of MatrixSymbol

### PR DESCRIPTION
**Pull request by sylee957**
#### Release Notes

- matrices
  - `MatrixSymbol` will store Str in its first argument.

### CODE REVIEW

File: sympy/printing/tests/test_tree.py [hunk 1]
```diff
@@ -2,178 +2,13 @@
...
-        '    negative: False\n',
-        '    noninteger: False\n',
-        '    nonnegative: True\n',
-        '    rational: True\n',
-        '    real: True\n',
-        '    transcendental: False\n'
-    ]
-
-    assert tree(A + B) == "".join(test_str)
+    # XXX After making the _assumptions cache deterministic,
```
Comment (eric-wieser) [inline sympy/printing/tests/test_tree.py:9]: Why did this file change? This seems unrelated to `MatrixSymbol`

Comment (sylee957) [reply sympy/printing/tests/test_tree.py:9]: This uses matrix symbols for test in printing.
And I’d find it painful to change every assumptions cache and its order even if they are generated in randomized way

Comment (eric-wieser) [reply sympy/printing/tests/test_tree.py:9]: But why do it in this PR? The test is already marked `XFAIL`.

File: sympy/printing/tests/test_tree.py [hunk 2]
```diff
@@ -2,178 +2,13 @@
...
-        '  nonzero: False\n',
-        '  odd: False\n',
-        '  positive: False\n',
-        '  prime: False\n',
-        '  rational: False\n',
-        '  real: False\n',
-        '  transcendental: False\n',
-        '  zero: False\n',
-        '  +-Symbol: B\n',
-        '  | commutative: True\n',
```
Comment (eric-wieser) [inline sympy/printing/tests/test_tree.py:135]: I assume all that needs changing here is:
```
        '  +-Str: B\n',
```

File: sympy/printing/tests/test_tree.py (hunk 1 above)
Comment (sylee957) [reply sympy/printing/tests/test_tree.py:9]: I wanted to make that this test should be run with dry run.

Comment (sylee957) [reply sympy/printing/tests/test_tree.py:9]: Or this can just be changed to test against actual string output if it doesn’t print out messy assumptions again

**Comment by asmeurer** [comment]
I missed the introduction of `Str`. I don't see anything in the release notes about it.

File: sympy/printing/tests/test_tree.py (hunk 1 above)
Comment (sylee957) [reply sympy/printing/tests/test_tree.py:8]: I clearly see that there is no way to avoid printing all the unnecessary assumptions

Comment (eric-wieser) [reply sympy/printing/tests/test_tree.py:8]: Is the issue that this prints lots of assumptions about `Str` itself, which obviously mean nothing?

Comment (sylee957) [reply sympy/printing/tests/test_tree.py:8]: Yes, and I just think that it’s the flaw of the original design of the printer

Comment (eric-wieser) [reply sympy/printing/tests/test_tree.py:8]: That sounds like a flaw in `Str` to me - assuming that a string is real / commutative etc doesn't make any sense. Marking as unresolved in case @asmeurer has thoughts.

Comment (sylee957) [reply sympy/printing/tests/test_tree.py:8]: It isn't only a problem of Str having assumptions, but all other sympy objects because Basic inherits assumptions from Expr.
Are you just saying you don't like the design of Str?

Comment (asmeurer) [reply sympy/printing/tests/test_tree.py:8]: I still think we shouldn't have Str in the first place, but if we do have it, it makes sense to me that it should be a Basic, not Expr object.

Comment (asmeurer) [reply sympy/printing/tests/test_tree.py:8]: I guess it already is Basic only, but the problem is that the default assumptions live on Basic rather than Expr. I think @oscarbenjamin discussed this issue before. It probably won't be easy to fix, but it is unrelated to Str specifically.

Comment (oscarbenjamin) [reply sympy/printing/tests/test_tree.py:8]: There are two problems here:

1. The old assumptions are applied to all of Basic rather than just Expr. I have a PR #19474 which experiments with changing that. It would not be hard to finish off that PR. The only thing missing is a deprecation warning wrapper that allows the old assumptions to keep working on Basic but with a warning.

2. The `tree` function should not print assumptions for all objects but rather only for those objects for which the assumptions are part of the definition (`Symbol`, `AppliedUndef`, anything else?). It should also only print those assumptions that are part of that definition (`.assumptions0`).

**Comment by czgdp1807** [comment]
@sylee957 Any news on this?

**Comment by sylee957** [comment]
This needs progress in #19841 to resolve the failing tests
